autopush/endpoint.py

```diff
@@ -833,52 +1064,54 @@ def _deleteChannel(self, message, uaid, chid):
         message.unregister_channel(uaid, chid)
 
     def _deleteUaid(self, message, uaid, router):
-        message.delete_all_for_user(uaid)
-        router.drop_user(uaid)
+        message.delete_user(uaid)
+        if not router.drop_user(uaid):
+            raise ItemNotFound("UAID not found")
 
     @cyclone.web.asynchronous
-    def delete(self, combo):
+    def delete(self, router_type="", router_token="", uaid="", chid=""):
         """HTTP DELETE
 
         Invalidate a UAID (and all channels associated with it).
 
         """
-        combo = combo.strip("/")
-        if "/" in combo:
-            (uaid, chid) = combo.split("/", 2)
-        else:
-            uaid = combo
-            chid = None
-        # what is request
+        # what is the request
         if not self._validate_auth(uaid):
             return self._write_response(
                 401, 109, message="Invalid Authentication")
+        if router_type not in self.ap_settings.routers:
+            log.msg("Invalid parameters", **self._client_info())
+            return self._write_response(
+                400, 108, message="Invalid arguments")
+        router = self.ap_settings.routers[router_type]
+        (valid, router_token) = router.check_token(router_token)
+        if not valid:
+            newUrl = self._relocate(router_type, router_token, uaid, chid)
+            return self._write_response(
+                301, 0, "Location: %s" % newUrl,
+                headers={"Location": newUrl})
+
         message = self.ap_settings.message
         if chid:
             # mark channel as dead
             self.ap_settings.metrics.increment("updates.client.unregister",
                                                tags=self.base_tags())
             d = deferToThread(self._deleteChannel, message, uaid, chid)
+            d.addCallback(self._success)
+            d.addErrback(self._chid_not_found_err)
             d.addErrback(self._response_err)
             return d
         # nuke uaid
         d = deferToThread(self._deleteUaid, message, uaid,
```

Comment:
Why is message pulled out and passed to deleteUaid? deleteUaid can get to the table just as easily.